4단계 - 로또(수동)

ghtjr410 opened the pull request:
# 로또(수동) 과제 제출
안녕하세요!
로또(수동) 과제 제출합니다.
## 미션 진행 과정
1. 기능 추가 시 컴파일 에러 없이 점진적인 리팩터링 (자동화 스크립트)
2. 수동 로또 생성기 구현 (ManualBasedLottoGenerator)
3. 수동/자동 구매 개수 관리 구조 설계
4. 사용자 입력 예외 처리 및 재입력 로직 구현
5. 당첨 번호 + 보너스 번호 검증 흐름 개선
## 고민했던 지점
### 1. ManualLottoCount vs LottoCount 통합
- 처음에는 ManualLottoCount(0 이상)와 LottoCount(1 이상)를 분리했습니다. 
- 하지만 이 방향으로 가면 AutoLottoCount도 생기고, 비슷한 값 객체가 계속 확산되는 문제가 있었습니다.
- 생각해보니 "1개 이상 구매해야 한다"는 LottoCount의 책임이 아니라 LottoPurchaseAmount의 책임이었습니다. 
- 1000원 이상이면 자연스럽게 1개 이상이 보장된다고 판단했습니다.
- 결론적으로 LottoCount를 0 이상으로 변경하고, subtract() 연산을 추가해서 총/수동/자동 개수를 하나의 타입으로 표현했습니다.
### 2. 재입력 범위 설계
- 당첨 번호와 보너스 번호 입력에서, 보너스만 잘못 입력해도 전체를 다시 받을지 고민했습니다.
- WinningLotto 생성 시점에 "보너스가 당첨번호와 중복" 검증이 일어나는데, 한 묶음으로 재시도하면 UX가 불편합니다. 
- 그래서 당첨번호 → Lotto 생성까지 먼저 완료하고, 보너스 번호는 별도로 재시도하도록 분리했습니다.
- 이 과정에서 기존 Lotto.contains() 메서드를 활용해 중복 검증 책임을 자연스럽게 위임할 수 있었습니다.
## 배운 점
- 비슷한 개념이 확산될 시 통합 가능성을 먼저 검토, 정말 다른 책임일 때만 분리
- 범용 메서드와 도메인 맥락 메시지는 충돌할 수 있고, 검증과 연산을 분리하면 해결 가능
***
리뷰 잘 부탁드립니다!

ghtjr410 committed:
docs: 로또 4단계(수동) 문서 추가 및 README 링크 추가

README.md
- 4단계 문서 링크를 `./docs/04-lotto-manual.md`로 수정하여 잘못된 링크를 정상화

04-lotto-manual.md
- 로또 4단계(수동) 요구사항 문서 신규 추가
- 기능 요구사항, 프로그래밍 규칙, 체크리스트, 구현 기능 목록 상세 기술

ghtjr410 committed:
feat: 수동 로또 생성기 및 테스트 추가

ManualBasedLottoGenerator.java
- 문자열 입력을 받아 로또 번호 파싱 후 Lotto 생성하도록 수동 로또 생성기 구현
- 기본 생성자에서 LottoNumberParser 주입 방식 제공

ManualBasedLottoGeneratorTest.java
- 문자열 목록 입력 시 올바른 Lotto 리스트가 생성되는지 검증 테스트 추가

ghtjr410 committed:
docs: 수동 로또 생성기 구현 사항 문서에 추가

04-lotto-manual.md
- 수동 로또 생성기(ManualBasedLottoGenerator) 기능 요구사항 항목 추가
- 문자열 목록을 기반으로 로또 목록을 생성하는 기능 명시

ghtjr410 committed:
feat: 수동 로또 구매 개수 도메인 및 테스트 추가

ManualLottoCount.java
- 문자열 입력을 정수로 변환하여 수동 구매 개수 생성
- null·빈값 입력 시 예외 발생 처리

ManualLottoCountTest.java
- 정상 숫자 입력 시 생성 성공 테스트
- null·빈 문자열 입력 시 예외 발생 여부 검증

ghtjr410 committed:
feat: 수동 로또 개수 검증 로직 강화 및 테스트 추가

ManualLottoCount.java
- 생성자 블록에서 음수 값 검증 로직 추가
- 음수 입력 시 "수동 구매 개수는 0 이상이어야 합니다." 예외 발생

ManualLottoCountTest.java
- 음수 입력 시 예외 발생 여부 검증 테스트 추가

ghtjr410 committed:
docs: 수동 구매 개수 기능 요구사항 추가

04-lotto-manual.md
- ManualLottoCount 요구사항 항목 추가
  - 0 이상 정수 생성 조건 명시
  - 문자열 생성 지원 명시
  - 빈 문자열 및 음수 입력 시 예외 발생 명시

ghtjr410 committed:
refactor: LottoCount 입력 검증 개선 및 문자열 생성자 추가

LottoCount.java
- 문자열 기반 생성자 추가 및 빈 문자열 검증 로직 도입
- 최소값 기준을 1 → 0 이상으로 변경
- 음수 입력 시 예외 메시지 "구매 개수는 0이상이어야 합니다."로 통일

LottoCountTest.java
- 정상 생성 가능 범위를 0 이상으로 확장한 테스트 추가
- 문자열 입력 시 빈값 예외 발생 테스트 추가
- 음수 입력 시 예외 검증으로 테스트 수정

ghtjr410 committed:
refactor: LottoCount 차감 기능 추가 및 테스트 보완

LottoCount.java
- subtract 메서드 추가하여 다른 LottoCount와의 차감 기능 제공

LottoCountTest.java
- subtract 호출 시 정상적으로 값이 감소하는지 검증 테스트 추가

ghtjr410 committed:
refactor: ManualLottoCount 제거

ManualLottoCount.java
- 수동 구매 개수 도메인 클래스 삭제

ManualLottoCountTest.java
- 관련 테스트 코드 전체 삭제

ghtjr410 committed:
feat: 수동 로또 구매 흐름 추가 및 입출력 기능 확장

Application.java
- 전체 구매 수에서 수동/자동 개수 분리 로직 추가
- 수동 번호 입력 및 로또 생성 기능 연동
- 자동/수동 로또 병합 처리 메서드 추가
- 구매 개수 출력 방식 변경(수동/자동 구분 출력)

InputView.java
- 수동 구매 개수 입력 메서드 추가
- 수동 번호 리스트 입력 메서드 추가

ResultView.java
- 구매 개수 출력 형식을 변경하여 수동/자동 개수 명시

ghtjr410 committed:
refactor: 불필요한 구매 개수 출력 로직 제거

PurchasedLottos.java
- purchaseCountForDisplay 및 size 관련 메서드 제거
  (구매 수 출력 책임을 ResultView로 완전 이관)

PurchasedLottosTest.java
- purchaseCountForDisplay 검증 테스트 삭제

ghtjr410 committed:
refactor: LottoCount 차감 검증 강화 및 테스트 확장

LottoCount.java
- subtract 호출 시 차감 가능 여부 검증 추가
- isLessThan 기반의 validateSubtractable 메서드 도입
- 차감 불가 시 "차감할 수 없습니다." 예외 발생하도록 수정

LottoCountTest.java
- subtract 정상/동일 값 케이스에 대한 CSV 기반 테스트 추가
- subtract 시 초과 차감 예외 테스트 추가
- validateSubtractable 정상/예외 케이스 테스트 추가

ghtjr410 committed:
feat: 예외 발생 시 재입력 가능하도록 재시도 로직 추가

Application.java
- 구매 금액 입력을 `retryUntilSuccess`로 감싸 예외 발생 시 재입력 가능하도록 개선
- 수동 구매 개수 입력 시 총 구매 수 초과 여부 검증 및 재입력 처리 추가
- 수동 로또 생성 단계에도 재시도 로직 적용
- 입력 람다 처리용 `retryUntilSuccess` 유틸 메서드 추가
- manual/auto 입력 분리 메서드(`getManualLottoCount`, `getManualLottoNumbers`) 추가

ghtjr410 committed:
refactor: 입력/출력 static import 적용 및 보너스 번호 검증 로직 분리

Application.java
- InputView, ResultView 메서드들을 static import로 교체하여 코드 가독성 개선
- 구매 금액/수동 개수/수동 번호/당첨 번호/보너스 번호 입력에 재시도(retry) 로직 적용
- 수동 로또 번호 입력 흐름 간소화 (`getManualLottoNumbers` 제거)
- 당첨 번호 파싱 및 보너스 번호 중복 검증 로직(`validateBonusNotDuplicated`) 분리
- WinningLotto 생성 시 LottoNumber를 직접 전달하도록 수정

WinningLotto.java
- 생성자 시그니처 변경: 보너스 번호를 String → LottoNumber로 받도록 일원화

ghtjr410 committed:
refactor: 당첨 로또 생성 흐름 개선 및 중복 검증 로직 단순화

Application.java
- WinningLotto 생성 과정 분리: `generateWinningLotto` 메서드 추가
- 당첨 번호 파싱 후 바로 Lotto 객체로 생성
- 보너스 번호 검증 시 `Lotto`의 contains 사용하도록 개선
- 기존 List 기반 검증 로직을 Lotto 기반으로 단순화

ghtjr410 committed:
refactor: 메서드 네이밍 정리 및 당첨 로또 생성 흐름 개선

Application.java
- 입력/생성 메서드 네이밍을 create* 스타일로 통일하여 책임 명확화
- 구매 로또 생성 흐름을 createManualLottos, generateAutoLottos로 단순화
- 당첨 번호 생성 흐름을 createLottoForWinning, createBonusNumber로 개선
- 전체적으로 불필요한 임시 변수 제거 및 호출 구조 간결화

ghtjr410 committed:
refactor: 불필요한 생성자 제거

WinningLotto.java
- List<LottoNumber>를 입력받는 생성자 제거하여 Lotto 기반 생성 방식으로 일원화
- 기존 int 보너스 번호 생성자는 그대로 유지

ghtjr410 committed:
docs: 로또 수동 구매 및 WinningLotto 변경사항 문서 반영

04-lotto-manual.md
- PR 링크 추가
- LottoCount 요구사항을 0 이상 생성 및 차감 검증 중심으로 업데이트
- ManualLottoCount 제거에 따른 관련 항목 삭제
- PurchasedLottos의 구매 개수 출력 기능 제거 반영
- WinningLotto 생성 방식 변경(로또 + 보너스 번호)으로 요구사항 수정

javajigi commented on src/main/java/lotto/domain/ManualBasedLottoGenerator.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.List;
+import lotto.domain.Lotto;
+
+public class ManualBasedLottoGenerator {
```
👍 
단, utils 성격의 객체는 아닌 것으로 보여짐

javajigi commented on src/main/java/lotto/Application.java:
```diff
@@ -1,38 +1,100 @@
-import lotto.view.ResultView;
+import lotto.utils.ManualBasedLottoGenerator;
 
 public class Application {
```
Application이 담당하는 책임이 너무 많아진 느낌
Lotto를 생성을 담당하는 책임을 별도의 객체로 분리해 보면 어떨까?
분리할 때 다음 피드백 참고해본다.

3단계에서 4단계로 요구사항이 변경될 때 로또를 생성하는 부분의 요구사항만 변경됐다.
로또를 생성하는 부분을 다음과 같은 구조의 인터페이스로 분리해 보는 연습을 해보면 어떨까?
이와 같이 인터페이스로 구현했을 때의 잇점에 대해 고민해 보는 시간을 가져본다.

```
public interface LottosGenerator {
    Lottos generate();
}
```

javajigi reviewed (changes requested):
다른 객체의 영향을 최소화하면서 4단계 미션 👍 
특별히 피드백할 것은 없는데요.
로또 생성과 관련해 추가로 도전해 봤으면 하는 부분이 있어 피드백 남겨 봅니다.
이 피드백 반영하면서 "수강신청 - 레거시 코드 리팩터링" 미션의 1단계 병행해볼 것을 추천합니다.

ghtjr410 committed:
refactor: 로또 생성 및 파싱 관련 클래스 패키지 구조 개선

ghtjr410 committed:
refactor: PurchasedLottos 도메인 명칭을 Lottos로 단순화

Application.java
- PurchasedLottos 타입을 Lottos로 변경
- 관련 메서드 반환 타입 및 생성 로직 수정

Lottos.java
- PurchasedLottos → Lottos로 클래스명 변경
- 생성자 및 레코드 선언부 명칭 일관성 개선

ResultView.java
- 구매 로또 출력 메서드 파라미터 타입을 Lottos로 변경

LottosTest.java
- 테스트 클래스 및 테스트 코드에서 Lottos 명칭으로 변경

ghtjr410 committed:
refactor: 도메인 관련 테스트 패키지를 domain으로 이동

AutoBasedLottoGeneratorTest.java
- 테스트 패키지를 utils에서 domain으로 변경
- 동일 패키지로 인한 불필요한 import 제거

LottoNumberParserTest.java
- 테스트 패키지를 utils에서 domain으로 변경
- 중복 import 제거로 코드 정리

ManualBasedLottoGeneratorTest.java
- 테스트 패키지를 utils에서 domain으로 변경
- 불필요한 도메인 import 제거

ghtjr410 committed:
refactor: 로또 생성 책임을 도메인으로 이동하고 생성기 인터페이스 도입

Application.java
- 수동/자동 로또 생성을 Lottos 단위로 처리하도록 변경
- merge 로직을 Application에서 제거하고 Lottos로 위임

AutoBasedLottoGenerator.java
- LottosGenerator 인터페이스 구현
- 생성 시 LottoCount를 주입받도록 생성자 구조 변경
- generate 결과를 Lottos로 반환하도록 수정

ManualBasedLottoGenerator.java
- LottosGenerator 인터페이스 구현
- 수동 입력값을 생성자에서 주입받도록 변경
- generate 결과를 Lottos로 반환

Lottos.java
- Lottos 병합 책임을 가지는 merge 메서드 추가
- size 메서드 추가로 컬렉션 역할 명확화

LottosGenerator.java
- 로또 생성 책임을 추상화한 인터페이스 추가

AutoBasedLottoGeneratorTest.java
- 변경된 생성자 및 반환 타입에 맞게 테스트 수정

ManualBasedLottoGeneratorTest.java
- generate 결과를 Lottos 기준으로 검증하도록 수정

LottosTest.java
- merge 기능에 대한 테스트 추가

ghtjr410 committed:
refactor: Application의 로또 생성 및 당첨 로직 간소화

Application.java
- 자동 로또 생성 메서드명을 generateAutoLottos → createAutoLottos로 변경해 네이밍 일관성 개선
- 당첨 로또 생성 관련 메서드들을 하나의 람다로 통합하여 로직 단순화
- 불필요한 import(List) 및 중간 메서드 제거로 코드 가독성 향상

ghtjr410 committed:
refactor: LottoGenerator 명명 일관성 개선 및 복수 개념 명확화

ghtjr410 committed:
feat: 수동/자동 로또 생성을 CombinedLottosGenerator로 통합

Application.java
- 수동/자동 로또 생성 로직을 CombinedLottosGenerator 사용 방식으로 리팩터링
- 개별 Manual/Auto 생성 메서드 제거 및 생성 책임 위임

CombinedLottosGenerator.java
- 수동/자동 LottosGenerator를 조합해 하나의 Lottos를 생성하는 Generator 추가
- 구매 금액과 수동 입력을 기반으로 Generator 구성을 캡슐화

CombinedLottosGeneratorTest.java
- 여러 Generator를 조합해 로또가 정상 생성되는지 검증하는 테스트 추가

ghtjr410 committed:
docs: 로또 생성 구조 변경에 따른 문서 업데이트

04-lotto-manual.md
- Lottos 도메인 역할 및 책임 명세 추가
- PurchasedLottos 제거에 따른 문서 정리
- LottosGenerator 인터페이스 및 구현체 구조 설명 추가
- Auto/Manual/Combined LottosGenerator 개념과 책임을 문서에 반영

ghtjr410 commented:
# 피드백 반영
## 변경사항
### 1. 패키지 이동
- `utils` -> `domain` 패키지로 Generator 클래스들 이동
- Generator들은 단순 유틸리티가 아닌 도메인 로직을 담당하므로 domain이 적절
### 2. LottosGenerator 인터페이스 도입
``` java
public interface LottosGenerator {
    Lottos generate();
}
```
- `AutoBasedLottosGenerator`, `ManualBasedLottosGenerator`가 구현
### 3. CombinedLottosGenerator 추가 (CompositePattern)
- 수동/자동 개수 계산 및 Generator 조합 책임을 Application에서 분리
- 여러 Generator를 조합하여 하나의 Lottos 반환
### 4. PurchasedLottos -> Lottos 변경
- `merge()` 메서드 추가로 로또 목록 병합 가능
### 5. Application 책임 축소
- 로또 생성 관련 로직을 CombinedLottosGenerator로 위임
- 불필요한 private 메서드 제거 (validateBonusNotDuplicated 등)
## 고민했던 지점
### 인터페이스로 구현했을 때의 장점
- OCP(개방-폐쇄 원칙) 달성
- 새로운 로또 생성 방식이 추가되어도 기존 코드 수정 없이 구현체만 추가하면 됩니다.
- 테스트 용이성: 인터페이스 덕분에 람다로 간단히 Mock 가능합니다.
``` java
LottosGenerator mockGen = () -> new Lottos(new Lotto(1,2,3,4,5,6));
``` 
### Composite Pattern 적용
처음에는 manual.merge(auto) 한 줄이면 충분하다고 생각했습니다. 
하지만 적용해보니 Application의 책임이 명확하게 줄어들었고
나중에 3개 이상의 Generator를 조합해야 할 때도 동일한 구조로 확장 가능하다는 점을 체감했습니다.
***
이번에도 리뷰 잘 부탁드립니다!

javajigi reviewed (approved):
피드백 반영 너무 잘 했네요. 💯 
특히 인터페이스 추가한 후에 CombinedLottosGenerator 까지 확장하는 것이 쉽지 않은데요.
특별히 피드백하지 않아도 이 단계까지 접근한 점이 인상적이네요.
지금까지 로또 미션 진행하느라 넘 수고했어요.
이 정도로 마무리할께요.
